crates/db/store-sled/src/ol/db.rs

```diff
@@ -0,0 +1,139 @@
+use strata_db_types::{
+    DbError, DbResult,
+    traits::{BlockStatus, OLBlockDatabase},
+};
+use strata_identifiers::{OLBlockCommitment, OLBlockId};
+use strata_ol_chain_types_new::OLBlock;
+
+use super::schemas::{OLBlockHeightSchema, OLBlockSchema, OLBlockStatusSchema};
+use crate::{
+    define_sled_database,
+    utils::{first, to_db_error},
+};
+
+define_sled_database!(
+    pub struct OLBlockDBSled {
+        blk_tree: OLBlockSchema,
+        blk_status_tree: OLBlockStatusSchema,
+        blk_height_tree: OLBlockHeightSchema,
+    }
+);
+
+impl OLBlockDatabase for OLBlockDBSled {
+    fn put_block_data(&self, commitment: OLBlockCommitment, block: OLBlock) -> DbResult<()> {
+        let block_id = *commitment.blkid();
+        let slot = commitment.slot();
+
+        self.config
+            .with_retry(
+                (&self.blk_tree, &self.blk_status_tree, &self.blk_height_tree),
+                |(bt, bst, bht)| {
+                    let mut blocks_at_slot = bht.get(&slot)?.unwrap_or(Vec::new());
+                    if !blocks_at_slot.contains(&block_id) {
+                        blocks_at_slot.push(block_id);
+                        bht.insert(&slot, &blocks_at_slot)?;
+
+                        // Only set status to Unchecked for new blocks
+                        // This preserves Valid/Invalid status if block is re-inserted
+                        bst.insert(&commitment, &BlockStatus::Unchecked)?;
+                    }
+
+                    bt.insert(&commitment, &block)?;
+                    Ok(())
+                },
+            )
+            .map_err(to_db_error)?;
+        Ok(())
+    }
+
+    fn get_block_data(&self, commitment: OLBlockCommitment) -> DbResult<Option<OLBlock>> {
+        Ok(self.blk_tree.get(&commitment)?)
+    }
+
+    fn del_block_data(&self, commitment: OLBlockCommitment) -> DbResult<bool> {
+        let block_id = *commitment.blkid();
+        let slot = commitment.slot();
+
+        self.config
+            .with_retry(
+                (&self.blk_tree, &self.blk_status_tree, &self.blk_height_tree),
+                |(bt, bst, bht)| {
+                    // Check if block exists
+                    if bt.get(&commitment)?.is_none() {
+                        return Ok(false);
+                    }
+
+                    let mut blocks_at_slot = bht.get(&slot)?.unwrap_or(Vec::new());
+                    blocks_at_slot.retain(|&bid| bid != block_id);
+
+                    bt.remove(&commitment)?;
+                    bst.remove(&commitment)?;
+                    bht.insert(&slot, &blocks_at_slot)?;
+                    Ok(true)
+                },
+            )
+            .map_err(to_db_error)
+    }
+
+    fn set_block_status(&self, commitment: OLBlockCommitment, status: BlockStatus) -> DbResult<()> {
+        // Check if block exists before setting status
+        if self.get_block_data(commitment)?.is_none() {
```

Comment:
Should we error out here? Or at least change the return value to indicate whether the status update succeeded or not. Because now we won't know whether we were actually able to update the status or not.